javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
I have also managed to write a test case that demonstrates the necessity of using `volatile` on the field of the double-checked locking mechanism. The test works on Java 6, 7 and 8. I managed to increase the chances of the concurrency problem being detected by introducing a _busy wait_ into the testing threads, acting as a live synchronization mechanism. The entire code is available [here](http://pastebin.com/P0hnVX4i).

It seems that Java 6 has different runtime behavior and it is easier to demonstrate the problem there.